### DISCUSSION TITLE
Incremental queries on associated objects

### DISCUSSION THREAD

**Discussion by mattparlane** [mongodb/mongoid#5737] [General]
Hi!

I was trying to do an incremental `or` query on an associated object, and the resulting query was not what I expected, and I wondered if it was intentional.

```
band.
  members.
  or(active: true).
  or(retired: true)

=> #<Mongoid::Criteria
  selector: {"$or"=>[{"band_id"=>BSON::ObjectId('652f51ec236b7a3e02183c0c')}, {"active"=>true}, {"retired"=>true}]}
  options:  {}
  class:    Member
  embedded: false>
```

I was expecting the `band_id` criteria to be added to each of the array elements. Is there any way to do what I'm looking for? Or would this be considered a bug?

Thanks!

**Comment by comandeo** [comment]
Hi @mattparlane , thank you for reporting this. I've added a ticket to our tracking system here, if you'd like to follow along: https://jira.mongodb.org/browse/MONGOID-5697

**Comment by jamis** [comment]
Hello @mattparlane, I spent some time looking into this. Firstly, I just want to point out that the example you gave would--even in the absense of existing bugs in Mongoid--probably not return the results you are looking for. You're asking for "all members of `band`, OR those that are `active`, OR those that are `retired`", and Mongoid is obligingly returning precisely that query. What I suspect you intended was to query for "all members of `band` WHERE `active` is true, OR all members of `band` WHERE `retired` is true".

The magic incantation that gives you that query is:

```ruby
band.
  members.
  where(active: true).or(
    band.
      members.
      where(retired: true)
  )
```

This mirrors the same semantics as ActiveRecord's `#or` clause, BTW.

Now, one issue here is that Mongoid lets you pass a query hash to the `#or` clause. This is a major difference between Mongoid's API and ActiveRecord's, since AR's API *requires* you to pass a *relation* there. If you tried to pass a Hash of conditions to `#or` in ActiveRecord, it would raise an error.

Now, whether Mongoid should support that same constraint or not, I'm not sure. I'll talk to the team about it, and perhaps we'll wind up prioritizing some changes to Mongoid so that the `#or` clause doesn't permit surprising queries; but as it would be a potentially breaking change, it will require some care to implement and schedule.

For now, hopefully the query I mentioned above will do the job for you.

**Comment by mattparlane** [comment]
Hey @jamis, thanks for getting back to me.

Totally understand what you're saying, and thanks for the query. I'll leave it with you guys as to where to take it from here.

Thanks again.
